core/preconditioner/schwarz.cpp

```diff
@@ -0,0 +1,208 @@
+
+        system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(), r.get());
+        auto restricted_r = Vector::create(
+            exec,
+            gko::dim<2>(restrict_op->get_size()[0], dense_x->get_size()[1]));
+        restrict_op->apply(r.get(), restricted_r.get());
+        auto restricted_e = Vector::create(
+            exec,
+            gko::dim<2>(restrict_op->get_size()[0], dense_x->get_size()[1]));
+        restricted_e->fill(zero<ValueType>());
+        parameters_.coarse_solver.coarse_solver_->apply(restricted_r.get(),
+                                                        restricted_e.get());
+        prolong_op->apply(one_op.get(), restricted_e.get(), one_op.get(),
+                          dense_x);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::apply_impl(const LinOp* alpha,
+                                               const LinOp* b,
+                                               const LinOp* beta,
+                                               LinOp* x) const
+{
+    precision_dispatch_real_complex<ValueType>(
+        [this](auto dense_alpha, const auto dense_b, auto dense_beta,
+               auto dense_x) {
+            size_type offset = 0;
+            // TODO Replace with BlockApprox
+            for (size_type i = 0; i < this->num_subdomains_; ++i) {
+                size_type l_num_rows = parameters_.subdomain_sizes[i];
+                auto rspan = gko::span(offset, offset + l_num_rows);
+                const auto b_view = dense_b->create_submatrix(
+                    rspan, gko::span(0, dense_b->get_size()[1]));
+                auto x_view = dense_x->create_submatrix(
+                    rspan, gko::span(0, dense_x->get_size()[1]));
+                this->subdomain_solvers_[i]->apply(dense_alpha, b_view.get(),
+                                                   dense_beta, x_view.get());
+                offset += l_num_rows;
+            }
+        },
+        alpha, b, beta, x);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::write(mat_data& data) const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Schwarz<ValueType, IndexType>::transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Schwarz<ValueType, IndexType>::conj_transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::generate(const LinOp* system_matrix,
+                                             bool skip_sorting)
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    using csr_type = matrix::Csr<ValueType, IndexType>;
+    const auto exec = this->get_executor();
+    auto csr_mtx =
+        convert_to_with_sorting<csr_type>(exec, system_matrix, skip_sorting);
```

Comment:
I wonder, in the distributed case do you get the `system_matrix` as something of the type `Matrix` (distributed) or do you get directly the local matrix which you can convert? Is that natural (no user involvement) in terms of interface? I guess that will depend on the distributed solver implementations as well.